services/src/contexts/migrations/database_migration.rs

```diff
@@ -0,0 +1,300 @@
+use async_trait::async_trait;
+use bb8_postgres::{bb8::PooledConnection, PostgresConnectionManager};
+use log::info;
+use snafu::ensure;
+use tokio_postgres::{
+    error::SqlState,
+    tls::{MakeTlsConnect, TlsConnect},
+    Socket, Transaction,
+};
+
+use crate::error::{Result, UnexpectedDatabaseVersionDuringMigration};
+
+pub type DatabaseVersion = String;
+
+/// The logic for migrating the database from one version to another.
+#[async_trait]
+pub trait Migration: Send + Sync {
+    /// The previous version of the database. `None` if this is the first migration.
+    /// The database must be in this version in order for the migration to be applied.
+    fn prev_version(&self) -> Option<DatabaseVersion>;
+
+    /// The new version of the database after applying this migration.
+    fn version(&self) -> DatabaseVersion;
```

Comment:
sure, but that still doesn't mean that only because the version of a migration is larger then the version of another migration, that it can be applied. It could be that there are other versions in between two versions.

We could make version an integer to avoid this. But that removes a lot of flexiblity, like linking the version so the major release, or adding some comment that gives some context.

Reply:
Ok, but we shouldn't point to the migration directly? Like a backward-linked list? Then we don't need the array? What is your take on this?